// react-layout-justified/dist/box.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Box = function (_Component) {
	_inherits(Box, _Component);

	function Box(props) {
		_classCallCheck(this, Box);

		var _this = _possibleConstructorReturn(this, (Box.__proto__ || Object.getPrototypeOf(Box)).call(this, props));

		_this.onBind();
		return _this;
	}

	_createClass(Box, [{
		key: 'onBind',
		value: function onBind() {
			this.onCreate = this.onCreate.bind(this);
			this.render = this.render.bind(this);
		}
	}, {
		key: 'onCreate',
		value: function onCreate() {}
	}, {
		key: 'render',
		value: function render() {
			var _props = this.props;
			var box = _props.box;
			var className = _props.className;
			var onBoxClicked = _props.onBoxClicked;
			var index = _props.index;
			var children = _props.children;

			var styles = _lodash2.default.pick(box, ['width', 'height', 'top', 'left']);
			styles.position = 'absolute';

			return _react2.default.createElement(
				'div',
				{
					onClick: onBoxClicked ? onBoxClicked.bind(null, box, index) : null,
					style: styles, className: className
				},
				children
			);
		}
	}]);

	return Box;
}(_react.Component);

Box.propTypes = {
	box: _react2.default.PropTypes.shape({
		width: _react2.default.PropTypes.number.isRequired,
		height: _react2.default.PropTypes.number.isRequired,
		top: _react2.default.PropTypes.number.isRequired,
		left: _react2.default.PropTypes.number.isRequired
	}).isRequired,
	className: _react2.default.PropTypes.string,
	index: _react2.default.PropTypes.number.isRequired,
	onBoxClicked: _react2.default.PropTypes.func
};
exports.default = Box;
