// react-layout-justified/dist/reactJustified.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _justifiedLayout = require('justified-layout');

var _justifiedLayout2 = _interopRequireDefault(_justifiedLayout);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _box = require('./box');

var _box2 = _interopRequireDefault(_box);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }
// import cn from 'classnames';


var ReactJustified = function (_Component) {
	_inherits(ReactJustified, _Component);

	function ReactJustified(props) {
		_classCallCheck(this, ReactJustified);

		var _this = _possibleConstructorReturn(this, (ReactJustified.__proto__ || Object.getPrototypeOf(ReactJustified)).call(this, props));

		_this.onBind();
		_this.init();
		return _this;
	}

	_createClass(ReactJustified, [{
		key: 'onBind',
		value: function onBind() {
			this.init = this.init.bind(this);
			this.render = this.render.bind(this);
			this.getGeometry = this.getGeometry.bind(this);
			this.generateBoxStyles = this.generateBoxStyles.bind(this);
			this.generateContainerStyles = this.generateContainerStyles.bind(this);
			this.generateBoxes = this.generateBoxes.bind(this);
		}
	}, {
		key: 'init',
		value: function init() {}
	}, {
		key: 'getGeometry',
		value: function getGeometry(className) {
			var options = this.props.options;
			return (0, _justifiedLayout2.default)(options.boxes, options);
		}
	}, {
		key: 'generateContainerStyles',
		value: function generateContainerStyles(geometry) {
			var containerWidth = this.props.options.containerWidth;

			return {
				height: geometry.containerHeight,
				width: containerWidth,
				position: 'relative'
			};
		}
	}, {
		key: 'generateBoxes',
		value: function generateBoxes(geometry) {
			var boxes = geometry.boxes;
			var _props = this.props;
			var boxInnerElements = _props.boxInnerElements;
			var onBoxClicked = _props.onBoxClicked;
			var boxClass = _props.boxClass;
			var widowCount = _props.widowCount;

			var boxesCount = boxes.length;
			var toDisplayCount = boxesCount - widowCount;

			return _lodash2.default.map(boxes, function (boxInfo, index) {
				if (toDisplayCount > index) return null;

				return _react2.default.createElement(
					_box2.default,
					{
						key: index,
						index: index,
						box: boxInfo,
						className: boxClass,
						onBoxClicked: onBoxClicked
					},
					boxInnerElements ? boxInnerElements[index] : null
				);
			});
		}
	}, {
		key: 'generateBoxStyles',
		value: function generateBoxStyles(boxInfo) {
			return {
				width: boxInfo.width,
				height: boxInfo.height,
				left: boxInfo.left
			};
		}
	}, {
		key: 'render',
		value: function render() {
			var geometry = this.getGeometry();
			var containerStyles = this.generateContainerStyles(geometry);
			var className = this.props.containerClass;

			return _react2.default.createElement(
				'div',
				{ className: className, style: containerStyles },
				this.generateBoxes(geometry)
			);
		}
	}]);

	return ReactJustified;
}(_react.Component);

ReactJustified.defaultProps = {
	options: {
		containerWidth: 1060,
		containerPadding: 10,
		boxSpacing: 10,
		targetRowHeight: 320,
		targetRowHeightTolerance: 0.25,
		maxNumRows: Number.POSITIVE_INFINITY,
		forceAspectRatio: false,
		showWidows: true,
		fullWidthBreakoutRowCadence: false
	},
	containerClass: 'justified-layout-container',
	boxClass: 'justified-layout-box'
};
ReactJustified.propTypes = {

	// Component specific props
	containerClass: _react2.default.PropTypes.oneOfType([_react2.default.PropTypes.string, _react2.default.PropTypes.arrayOf(_react2.default.PropTypes.string)]),
	boxClass: _react2.default.PropTypes.oneOfType([_react2.default.PropTypes.string, _react2.default.PropTypes.arrayOf(_react2.default.PropTypes.string)]),
	boxInnerElements: _react2.default.PropTypes.arrayOf(_react2.default.PropTypes.element),
	onBoxClicked: _react2.default.PropTypes.func,
	onContainerClicked: _react2.default.PropTypes.func,

	// Justified layout specific
	options: _react2.default.PropTypes.shape({
		boxes: _react2.default.PropTypes.arrayOf(_react2.default.PropTypes.oneOfType([_react2.default.PropTypes.number.isRequired, _react2.default.PropTypes.shape({
			width: _react2.default.PropTypes.number.isRequired,
			height: _react2.default.PropTypes.number.isRequired
		})])).isRequired,
		containerWidth: _react2.default.PropTypes.number,
		containerPadding: _react2.default.PropTypes.oneOfType([_react2.default.PropTypes.number, _react2.default.PropTypes.shape({
			top: _react2.default.PropTypes.number.isRequired,
			right: _react2.default.PropTypes.number.isRequired,
			bottom: _react2.default.PropTypes.number.isRequired,
			left: _react2.default.PropTypes.number.isRequired
		})]),
		boxSpacing: _react2.default.PropTypes.oneOfType([_react2.default.PropTypes.number, _react2.default.PropTypes.shape({
			horizontal: _react2.default.PropTypes.number.isRequired,
			vertical: _react2.default.PropTypes.number.isRequired
		})]),
		targetRowHeight: _react2.default.PropTypes.number,
		targetRowHeightTolerance: function targetRowHeightTolerance(props, propName, componentName) {
			var targetRowHeightToleranceValue = props[propName];

			if (!targetRowHeightToleranceValue) return null;

			var errorMessage = void 0;

			if (typeof targetRowHeightToleranceValue !== 'number') {
				errorMessage = 'Props validation error in ' + componentName + ' component.\n\t\t\t        Expected ' + propName + ' to have type number, but got ' + (typeof prop === 'undefined' ? 'undefined' : _typeof(prop));
			}

			if (targetRowHeightToleranceValue > 1 || targetRowHeightToleranceValue < 0) {
				errorMessage = 'Props validation error in ' + componentName + ' component.\n\t\t\t\t\tExpected ' + propName + ' to be in range of 0 < ' + propName + ' < 1, but got ' + targetRowHeightToleranceValue + ' value';
			}

			if (errorMessage) return new Error(errorMessage);

			return null;
		},
		maxNumRows: _react2.default.PropTypes.number,
		forceAspectRatio: _react2.default.PropTypes.oneOfType([_react2.default.PropTypes.bool, _react2.default.PropTypes.number]),
		showWidows: _react2.default.PropTypes.bool,
		fullWidthBreakoutRowCadence: _react2.default.PropTypes.oneOfType([_react2.default.PropTypes.bool, _react2.default.PropTypes.number])
	}).isRequired
};
exports.default = ReactJustified;
