// react-layout-justified/lib/box.js
import React, { Component } from 'react';
import _ from 'lodash';

class Box extends Component {
	static propTypes = {
		box: React.PropTypes.shape({
			width: React.PropTypes.number.isRequired,
			height: React.PropTypes.number.isRequired,
			top: React.PropTypes.number.isRequired,
			left: React.PropTypes.number.isRequired,
		}).isRequired,
		className: React.PropTypes.string,
		index: React.PropTypes.number.isRequired,
		onBoxClicked: React.PropTypes.func,
	};

	constructor(props) {
		super(props);

		this.onBind();
	}

	onBind() {
		this.onCreate = this.onCreate.bind(this);
		this.render = this.render.bind(this);
	}

	onCreate() {

	}

	render() {
		const { box, className, onBoxClicked, index, children } = this.props;
		const styles = _.pick(box, [
			'width',
			'height',
			'top',
			'left',
		]);
		styles.position = 'absolute';

		return (
			<div
				onClick={ onBoxClicked ? onBoxClicked.bind(null, box, index) : null }
				style={ styles } className={ className }
			>
				{ children }
			</div>
		);
	}
}

export default Box;
