// react-layout-justified/lib/reactJustified.js
import React, { Component } from 'react';
import justifiedLayout from 'justified-layout';
// import cn from 'classnames';
import _ from 'lodash';

import Box from './box';

class ReactJustified extends Component {
	static defaultProps = {
		options: {
			containerWidth: 1060,
			containerPadding: 10,
			boxSpacing: 10,
			targetRowHeight: 320,
			targetRowHeightTolerance: 0.25,
			maxNumRows: Number.POSITIVE_INFINITY,
			forceAspectRatio: false,
			showWidows: true,
			fullWidthBreakoutRowCadence: false,
		},
		containerClass: 'justified-layout-container',
		boxClass: 'justified-layout-box',
	};

	static propTypes = {

		// Component specific props
		containerClass: React.PropTypes.oneOfType([
			React.PropTypes.string,
			React.PropTypes.arrayOf(React.PropTypes.string),
		]),
		boxClass: React.PropTypes.oneOfType([
			React.PropTypes.string,
			React.PropTypes.arrayOf(React.PropTypes.string),
		]),
		boxInnerElements: React.PropTypes.arrayOf(React.PropTypes.element),
		onBoxClicked: React.PropTypes.func,
		onContainerClicked: React.PropTypes.func,

		// Justified layout specific
		options: React.PropTypes.shape({
			boxes: React.PropTypes.arrayOf(React.PropTypes.oneOfType([
				React.PropTypes.number.isRequired,
				React.PropTypes.shape({
					width: React.PropTypes.number.isRequired,
					height: React.PropTypes.number.isRequired,
				}),
			])).isRequired,
			containerWidth: React.PropTypes.number,
			containerPadding: React.PropTypes.oneOfType([
				React.PropTypes.number,
				React.PropTypes.shape({
					top: React.PropTypes.number.isRequired,
					right: React.PropTypes.number.isRequired,
					bottom: React.PropTypes.number.isRequired,
					left: React.PropTypes.number.isRequired,
				}),
			]),
			boxSpacing: React.PropTypes.oneOfType([
				React.PropTypes.number,
				React.PropTypes.shape({
					horizontal: React.PropTypes.number.isRequired,
					vertical: React.PropTypes.number.isRequired,
				}),
			]),
			targetRowHeight: React.PropTypes.number,
			targetRowHeightTolerance: (props, propName, componentName) => {
				const targetRowHeightToleranceValue = props[propName];

				if (!targetRowHeightToleranceValue) return null;

				let errorMessage;

				if (typeof targetRowHeightToleranceValue !== 'number') {
					errorMessage = `Props validation error in ${componentName} component.
			        Expected ${propName} to have type number, but got ${ typeof prop }`;
				}

				if (targetRowHeightToleranceValue > 1 || targetRowHeightToleranceValue < 0) {
					errorMessage = `Props validation error in ${componentName} component.
					Expected ${propName} to be in range of 0 < ${propName} < 1, but got ${targetRowHeightToleranceValue} value`;
				}

				if (errorMessage) return new Error(errorMessage);

				return null;
			},
			maxNumRows: React.PropTypes.number,
			forceAspectRatio: React.PropTypes.oneOfType([
				React.PropTypes.bool,
				React.PropTypes.number,
			]),
			showWidows: React.PropTypes.bool,
			fullWidthBreakoutRowCadence: React.PropTypes.oneOfType([
				React.PropTypes.bool,
				React.PropTypes.number,
			]),
		}).isRequired,
	};

	constructor(props) {
		super(props);

		this.onBind();
		this.init();
	}

	onBind() {
		this.init = this.init.bind(this);
		this.render = this.render.bind(this);
		this.getGeometry = this.getGeometry.bind(this);
		this.generateBoxStyles = this.generateBoxStyles.bind(this);
		this.generateContainerStyles = this.generateContainerStyles.bind(this);
		this.generateBoxes = this.generateBoxes.bind(this);
	}

	init() {

	}

	getGeometry(className) {
		const options = this.props.options;
		return justifiedLayout(options.boxes, options);
	}

	generateContainerStyles(geometry) {
		const { containerWidth } = this.props.options;
		return {
			height: geometry.containerHeight,
			width: containerWidth,
			position: 'relative',
		};
	}

	generateBoxes(geometry) {
		const boxes = geometry.boxes;
		const { boxInnerElements, onBoxClicked, boxClass, widowCount } = this.props;
		const boxesCount = boxes.length;
		const toDisplayCount = boxesCount - widowCount;

		return _.map(boxes, (boxInfo, index) => {
			if (toDisplayCount > index) return null;
			
			return <Box
				key={ index }
				index={ index }
				box={ boxInfo }
				className={ boxClass }
				onBoxClicked={ onBoxClicked }
			>
				{ boxInnerElements ? boxInnerElements[index] : null }
			</Box>;
		});
	}

	generateBoxStyles(boxInfo) {
		return {
			width: boxInfo.width,
			height: boxInfo.height,
			left: boxInfo.left,
		};
	}

	render() {
		const geometry = this.getGeometry();
		const containerStyles = this.generateContainerStyles(geometry);
		const className = this.props.containerClass;

		return (
			<div className={ className } style={ containerStyles }>
				{ this.generateBoxes(geometry)}
			</div>
		);
	}
}

export default ReactJustified;
